app/models/hipster.rb

```diff
@@ -0,0 +1,45 @@
+class Hipster < ActiveRecord::Base
+  # Include default devise modules. Others available are:
+  # :confirmable, :lockable, :timeoutable and :omniauthable
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :trackable, :validatable, :omniauthable, :omniauth_providers => [:facebook]
+
+attr_accessor :login
+ # validate :validate_hipstername
+  #validates :email, uniqueness: true
+  #validates :hipstername, uniqueness: true
+  validates :password, confirmation: true
+  validates :password_confirmation, presence: true
+
+  has_many :hipstergrams, dependent: :destroy
+  has_many :like_before_cools
+
+#def validate_hipstername
+ # if Hipster.where(email: hipstername).exists?
+ #   errors.add(:hipstername, :invalid)
+ # end
+#end
+
+def self.from_omniauth(auth)
+  where(provider: auth.provider, uid: auth.uid).first_or_create! do |user|
+    user.email = auth.info.email
+    user.password = Devise.friendly_token[0,20]
+    user.name = auth.info.name   # assuming the user model has a name
+  end
+end
+
+
+def self.find_first_by_auth_conditions(warden_conditions)
+  conditions = warden_conditions.dup
+  if login = conditions.delete(:login)
+    where(conditions).where(["lower(hipstername) = :value OR lower(email) = :value", { :value => login.downcase }]).first
```

Comment:
Metrics/LineLength: Line is too long. [121/80] (https://github.com/bbatsov/ruby-style-guide#80-character-limits)<br>Style/HashSyntax: Use the new Ruby 1.9 hash syntax. (https://github.com/bbatsov/ruby-style-guide#hash-literals)